File: queue.py
"""
@author     : Rajan Khullar
@created    : 01/08/16
@updated    : 01/08/16
"""


class Node:

    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

    def __str__(self):
        return str(self.data)

    def has_next(self):
        return self.next is not None

    def has_prev(self):
        return self.prev is not None


class Queue:

    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None
        self.curr = None

    def __str__(self):
        self.curr = self.head
        if self.curr is None:
            return '[]'
        out = '['
        while self.curr.has_next():
            out += str(self.curr) + ', '
            self.curr = self.curr.next
        if self.curr == self.tail:
            out += str(self.curr) + ']'
        return out

    def empty(self):
        return self.head is None and self.tail is None

    def peek(self):
        return self.head.data

    def enq(self, item):
        node = Node(item)
        if self.tail is None:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.size += 1

    def deq(self):
        self.curr = self.head
        if self.curr is None:
            print('cannot deque empty queue')
            return False
        item = self.head.data
        if self.curr.has_next():
            self.curr = self.curr.next
            self.head = self.curr
            self.size -= 1
        else:
            self.head = None
            self.tail = None
            self.size = 0
        return item

File: Queue.java
public class Queue<T>
{
    private Node<T> head, tail;
    private int size;

    public Queue()
    {
        reset();
    }

    public Queue(T data)
    {
        reset(data);
    }

    public void reset()
    {
        head = null;
        tail = null;
        size = 0;
    }

    public void reset(T data)
    {
        head = new Node<>(data);
        tail = head;
        size = 1;
    }

    public boolean isEmpty()
    {
        return head == null && tail == null;
    }

    public int size()
    {
        return size;
    }

    public void enqueue(T data)
    {
        if(isEmpty())
            {reset(data); return;}
        Node<T> node = new Node<>(data);
        node.setPrev(tail);
        tail.setNext(node);
        tail = node;
        size++;
    }

    public T dequeue()
    {
        if(isEmpty())
            return null;
        T data = head.getData();
        if(size == 1)
            {reset(); return data;}
        Node<T> node = head.getNext();
        head.setNext(null);
        node.setPrev(null);
        head = node;
        size--;
        return data;
    }

    public void absorb(Queue<T> o)
    {
        while(!o.isEmpty())
            enqueue(o.dequeue());
    }

    public Node<T> getHead()
    {
        return head;
    }

    public Node<T> getTail()
    {
        return tail;
    }

    public String toString()
    {
        if(isEmpty()) return "[]";
        String s = "[";
        Node<T> node = head;
        while(node.hasNext())
        {
            s += node.toString() + ", ";
            node = node.getNext();
        }
        s += node.toString() + "]";
        return s;
    }
}
